### DISCUSSION TITLE
[4.0] Fix image handling in media manager

### DISCUSSION THREAD

**Discussion by N6REJ** [joomla/joomla-cms#36157] [General]
Image handling in tinymce is abissmal at best.  When you select an image you need to toggle the drop down for "advance" to get to insert then you need to select the image and hit the image icon to make adjustsments, which you can't do because the "style" field has broken editing ability.  
"Vertical spacing" & "Horizontal spacing" are mediocre at best.
The user should have full control over how they want the image laid out and simply saying "use classes" is not acceptable for a USER.
For a dev possibly but they shouldn't be required to jump through so many hoops....
![image](https://user-images.githubusercontent.com/1850089/143916076-9931c81c-6fe7-4c73-a2a5-51dbe25d39d2.png)

Reizing function is crippled...
you can't change from pixels to any other measurement so your images cant' be made responsive.  nor does the full image sizing value show.
![image](https://user-images.githubusercontent.com/1850089/143916470-f7a6fe0a-7af2-4ac3-967f-3658084a1dbd.png)

**WE CAN DO BETTER THEN THIS!!!!**

**Comment by brianteeman** [comment]
> Reizing function is crippled...

I cannot replicate your issue with the value not showing

![input](https://user-images.githubusercontent.com/1296369/143921917-4c7ea527-5f47-4330-b0c8-6ed245250381.gif)

px is the only valid unit of measurement for an image

> WE CAN DO BETTER THEN THIS!!!!

Glad to see that you feel so strongly about this that you are shouting. Looking forward to seeing some constructive proposals

**Reply by N6REJ** [reply]
> px is the only valid unit of measurement for an image

Not sure how you can justify that statement.  Surely you know thats false for responsive images.

I did discover an issue with FF dev but it seems the original statement of resizing value not showing fully is corrected in git.  I know she's using a "release" version.
![image](https://user-images.githubusercontent.com/1850089/144066344-47cb652a-b2bd-4ebd-8124-72ae82b5e907.png)

> Glad to see that you feel so strongly about this that you are shouting.

I do feel strongly about Joomla!  We are letting so many little things that should've shown up in testing slip thru that its frustrating and quite frankly embarassing.  When I have a designer tell me she can't possibly use 4 for clients and show me why it's frustrating.  It feels like more and more we're writing Joomla! for dev's not users.

> Looking forward to seeing some constructive proposals

I do what I can.  The last 6+ mo have been rough for me personally.  Not going to go into details here, JBS knows whats up.  Even just posting a bug and explaining why its a bug is "constructive".  It's just as valid as a pr for a typo.

**Comment by brianteeman** [comment]
> > px is the only valid unit of measurement for an image

> Not sure how you can justify that statement. Surely you know thats false for responsive images.

What other units of measurement are valid for an img

**Reply by N6REJ** [reply]
Don't play dumb Brian, your not that stupid

**Reply by brianteeman** [reply]
I genuinely am. I am not aware of any unit of measurement that is valid for an img other than the px unit

**Reply by N6REJ** [reply]
% are the most common.  This is the only way images can be responsive.  There is also "fill", "container", "Cover" which use another type of percentile sizing.
Lets say for example you always wanted the image to take up 25% of the top left of a "card", maybe its a thumbnail.  Well, the card would generally be some % of its parent, so if you have a static image size, its not going to resize and maintain ratios with static sizing.
If the container is fluid there's a very high chance you want the image fluid as well.
Technically you could also use vw and vh but that would be a special use case.
OH I forgot "auto" very important for responsive imagery

**Reply by brianteeman** [reply]
Those are css classes not units of measurement for an img

**Reply by N6REJ** [reply]
that doesn't invalidate the rest of the issue(s)

**Reply by brianteeman** [reply]
it does for me

**Reply by N6REJ** [reply]
so, the fact you can't control your image sizing or layout is just fine with you.. Ok.. very constructive.

**Reply by brianteeman** [reply]
of course you can.

**Reply by richard67** [reply]
@N6REJ With image and pixels Brian is right. Check here: https://www.w3schools.com/html/html_images.asp

> The `width` and `height` attributes always define the width and height of the image in pixels.

For responsive images you use the picture element: https://www.w3schools.com/html/html_images_picture.asp

**Reply by dgrammatiko** [reply]
The attributes `width` and `height` are unitless and used from the browser to calculate the aspect ratio so that the page doesn't jump around when an image is fetched. The values for these CAN ONLY be the actual values of the image or values to calculate the same aspect ratio. These attributes have nothing to do with the sizing of an image in the document for this you either use an inline style or, because it's 2022 (almost) a class.

There's nothing missing, as far as I understand all the provided inputs provided...

**Comment by N6REJ** [comment]
@richard67 @dgrammatiko ty for starting a constructive conversation.  I think in the debate over px or non px the main issue got sidetracked.
The issue with the px input box not being sized properly is fixed in dev so thats not an issue.. And being its a "resize" screen px are correct.
SO, having said all that, this video should clearly explain the real issue that we're currently seeing.
Please keep in mind I'm relaying the feedback that has been given to me and why J4 can't be used for the average client.

Youtube video explaining: https://youtu.be/yVjWx4erK4k

**Reply by dgrammatiko** [reply]
You do realise that the plugin is **NOT** some joomla thing but code directly from tinyMCE. I guess you have to open an issue to their repo, this has nothing to do with Joomla's code

**Reply by N6REJ** [reply]
I was under the impression we had modified it substantially

**Reply by dgrammatiko** [reply]
Nope, the only plugin that is Joomla's own code is the jdraganddrop (I wrote that) as you can see here: https://github.com/joomla/joomla-cms/tree/4.0-dev/build/media_source/plg_editors_tinymce/js/plugins

**Reply by N6REJ** [reply]
hmmm.. so how do we.. protect?... our image?  Do we just tell clients, they need to use a real editor like jce?

**Reply by N6REJ** [reply]
doubt it will do any good but submitted https://github.com/tinymce/tinymce/issues/7442

**Reply by dgrammatiko** [reply]
jce is also tinyMCE and an outdated one. Please make a list of things that you're missing and maybe someone will be able to do something about that

**Reply by N6REJ** [reply]
ok.  Let me talk to ali since she's the designer that has ... difficult... clients the most and get her full list.  This video is what was recently brought up.
The video shows the main thing but let me make it clearer whats going on.  
Ty

**Reply by dgrammatiko** [reply]
BTW if people are still using inline styles to do the styling they need to do some catch-up and some CSS reading...
Anyways if you get the list ping me

**Comment by AlisonAMG** [comment]
Hi Dimitris - The issue isn't me (integrator level) - we can learn and yes do the class thing. I'm talking about my clients who manage their own sites and even with training and reminders they still use WYSIWYG to get things done day to day. Making the transition from older versions of J into J4 it will be harder still as they are used to what they could do day to day. That may not be perfect but it is reality and even more so if they have managed to cobble the site together themselves and don't have someone training (or trying to - they also have habits and when rushing to get work done they just want it done). 

I have used JCE since forever on my sites but am taking the time to get comfy with the new Media Manager as I know many many hours have gone into making it. I've also been watching you talking about JCE using a very old Tiny version. 

Currently in J4 MM to add an image, right align, add spacing to the left you have to do these steps; use the CMS/Media tool to find the image and place it (current position left), select the image and click align right as if aligning text (current position now right), now select image again and go back to CMS/Media tool and add the left wrap spacing. This is just wrong with wrong sauce. 

Since the correct way is to use classes could a couple classes be added (radio or other easy visible option) for left/center/right align? 
With that in combo with the other option in advanced for spacing might do the trick while still moving them to classes.

**Reply by dgrammatiko** [reply]
Did you read the Bootstrap docs for positioning images? 
https://getbootstrap.com/docs/5.1/content/images/#aligning-images

Here's some images that could act as a guide:

![Screenshot 2021-12-04 at 20 09 53](https://user-images.githubusercontent.com/3889375/144721829-e3f7dcb1-8e01-4926-b529-fbe396fb00d6.png)


![Screenshot 2021-12-04 at 20 10 28](https://user-images.githubusercontent.com/3889375/144721841-b058433b-358f-4f99-b950-ff05dc2e4f5d.png)

![Screenshot 2021-12-04 at 20 11 14](https://user-images.githubusercontent.com/3889375/144721847-8d0b2f86-b46a-4044-ae10-9cefb389dacf.png)


![Screenshot 2021-12-04 at 20 13 33](https://user-images.githubusercontent.com/3889375/144721856-285748ba-a982-40a6-8333-9326397aade2.png)


So, as I'm quite naive, what exactly is missing here, apart that people using technology like it's 2006?

**Reply by dgrammatiko** [reply]
> Since the correct way is to use classes could a couple classes be added (radio or other easy visible option) for left/center/right align?

That is not as simple as it sounds. The classes are totally related to the front end CSS framework that the template is using. If Joomla adds classes for Bootstrap then other CSS frameworks (and thaus the templates using them) won't be able to correctly position the images. Actually this was the reason the class field was left as a text field so people could add any class... (I know probably not the answer you were clooking for)

**Reply by AlisonAMG** [reply]
You're right it isn't. That said, thank you for saying why. 

I'll carry on building and end users can suck it up.

**Reply by dgrammatiko** [reply]
> I'll carry on building and end users can suck it up.

As an integrator, you can provide pre-built templates for the content, @brianteeman has a great tutorial on this subject.

**Reply by AlisonAMG** [reply]
Thank you for your time. 

I can provide pre-builts (I'm loving the possibilities with the Stamp tool as Brian showed in his JUG NJ talk) but sometimes there are cases that their post doesn't fit a preordained style. 

Anyhow like I said, I am trying to present a point for those who typically never get a voice here and simply choose another CMS.

**Reply by brianteeman** [reply]
Not every piece of software will ever be suitable for everyone (and nor should it)

**Reply by AlisonAMG** [reply]
Fully agree Brian. I just remember the big flap at the forum for making a more friendly interface.
